### PR TITLE
[WIP][SPARK-35084][CORE] Spark 3: supporting "--packages" in  k8s cluster mode

### PR DESCRIPTION
**Pull request by ocworld**
### What changes were proposed in this pull request?
Supporting '--packages' in the k8s cluster mode

### Why are the changes needed?
In spark 3, '--packages' in the k8s cluster mode is not supported. I expected that managing dependencies by using packages like spark 2.

Spark 2.4.5

https://github.com/apache/spark/blob/v2.4.5/core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala

```scala
     if (!isMesosCluster && !isStandAloneCluster) {
      // Resolve maven dependencies if there are any and add classpath to jars. Add them to py-files
      // too for packages that include Python code
      val resolvedMavenCoordinates = DependencyUtils.resolveMavenDependencies(
        args.packagesExclusions, args.packages, args.repositories, args.ivyRepoPath,
        args.ivySettingsPath)
      
      if (!StringUtils.isBlank(resolvedMavenCoordinates)) {
        args.jars = mergeFileLists(args.jars, resolvedMavenCoordinates)
        if (args.isPython || isInternal(args.primaryResource)) {
          args.pyFiles = mergeFileLists(args.pyFiles, resolvedMavenCoordinates)
        }
      } 
      
      // install any R packages that may have been passed through --jars or --packages.
      // Spark Packages may contain R source code inside the jar.
      if (args.isR && !StringUtils.isBlank(args.jars)) {
        RPackageUtils.checkAndBuildRPackage(args.jars, printStream, args.verbose)
      }
    } 
 ```

Spark 3.0.2

https://github.com/apache/spark/blob/v3.0.2/core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala

```scala
       if (!StringUtils.isBlank(resolvedMavenCoordinates)) {
        // In K8s client mode, when in the driver, add resolved jars early as we might need
        // them at the submit time for artifact downloading.
        // For example we might use the dependencies for downloading
        // files from a Hadoop Compatible fs eg. S3. In this case the user might pass:
        // --packages com.amazonaws:aws-java-sdk:1.7.4:org.apache.hadoop:hadoop-aws:2.7.6
        if (isKubernetesClusterModeDriver) {
          val loader = getSubmitClassLoader(sparkConf)
          for (jar <- resolvedMavenCoordinates.split(",")) {
            addJarToClasspath(jar, loader)
          }
        } else if (isKubernetesCluster) {
          // We need this in K8s cluster mode so that we can upload local deps
          // via the k8s application, like in cluster mode driver
          childClasspath ++= resolvedMavenCoordinates.split(",")
        } else {
          args.jars = mergeFileLists(args.jars, resolvedMavenCoordinates)
          if (args.isPython || isInternal(args.primaryResource)) {
            args.pyFiles = mergeFileLists(args.pyFiles, resolvedMavenCoordinates)
          }
        }
      }
```

unlike spark2, in spark 3, jars are not added in any place.

### Does this PR introduce _any_ user-facing change?
Unlike spark 2, resolved jars are added not in cluster mode spark submit but in driver.

It's because in spark 3, the feature is added that is uploading jars with prefix "file://" to s3.
So, if resolved jars are added in spark submit, every jars from packages are uploading to s3! When I tested it, it is very bad experience to me.

### How was this patch tested?
In my k8s environment, i tested the code.

### CODE REVIEW

**Comment by ocworld** [comment]
@AhnLab-OSS

**Comment by ocworld** [comment]
Can I see logs about failure of build and test?

**Comment by dongjoon-hyun** [comment]
Thank you for your contribution, @ocworld .

**Comment by dongjoon-hyun** [comment]
ok to test

**Comment by dongjoon-hyun** [comment]
> Can I see logs about failure of build and test?

Apache Spark GitHub Action is running in your GitHub Account.

Screen Shot 2021-05-01 at 4 11 15 PM

If you click the failure link, you can see the error log in your repo. The current failure occurs because your master branch is too outdate. Please follow the guideline to sync your master branch.
- https://github.com/apache/spark/pull/32397/checks?check_run_id=2466811890
```
Unable to detect the workflow run for testing the changes in your PR.

If you did not enable GitHub Actions in your forked repository, please enable it. See also Disabling or limiting GitHub Actions for a repository for more details.
It is possible your branch is based on the old master branch in Apache Spark, please sync your branch to the latest master branch. For example as below:
git fetch upstream
git rebase upstream/master
git push origin YOUR_BRANCH --force
```

In addition, I also triggered Jenkins machine too. For those users, we have another test machine additionally.

**Comment by ocworld** [comment]
@dongjoon-hyun Thank you for your test and reply.
Though rebasing from upstream, checking build and test is failed now.
Jenkins build is also failed with logs.

```
Setting status of fe0a4b5625ffceaa6c81e9f63637214d08369642 to FAILURE with url https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/42657/ and message: 'Build finished. '
FileNotFoundException means that the credentials Jenkins is using is probably wrong. Or the user account does not have write access to the repo.
org.kohsuke.github.GHFileNotFoundException: https://api.github.com/repos/apache/spark/statuses/fe0a4b5625ffceaa6c81e9f63637214d08369642 {"message":"Not Found","documentation_url":"https://docs.github.com/rest/reference/repos#create-a-commit-status"}
	at org.kohsuke.github.GitHubClient.interpretApiError(GitHubClient.java:492)
	at org.kohsuke.github.GitHubClient.sendRequest(GitHubClient.java:420)
	at org.kohsuke.github.GitHubClient.sendRequest(GitHubClient.java:363)
	at org.kohsuke.github.Requester.fetch(Requester.java:74)
	at org.kohsuke.github.GHRepository.createCommitStatus(GHRepository.java:1906)
	at org.jenkinsci.plugins.ghprb.extensions.status.GhprbSimpleStatus.createCommitStatus(GhprbSimpleStatus.java:283)
	at org.jenkinsci.plugins.ghprb.extensions.status.GhprbSimpleStatus.onBuildComplete(GhprbSimpleStatus.java:241)
	at org.jenkinsci.plugins.ghprb.GhprbBuilds.onCompleted(GhprbBuilds.java:205)
	at org.jenkinsci.plugins.ghprb.GhprbBuildListener.onCompleted(GhprbBuildListener.java:28)
	at hudson.model.listeners.RunListener.fireCompleted(RunListener.java:209)
	at hudson.model.Run.execute(Run.java:1952)
	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
	at hudson.model.ResourceController.execute(ResourceController.java:97)
	at hudson.model.Executor.run(Executor.java:429)
Caused by: java.io.FileNotFoundException: https://api.github.com/repos/apache/spark/statuses/fe0a4b5625ffceaa6c81e9f63637214d08369642
	at sun.reflect.GeneratedConstructorAccessor1029.newInstance(Unknown Source)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at sun.net.www.protocol.http.HttpURLConnection$10.run(HttpURLConnection.java:1950)
	at sun.net.www.protocol.http.HttpURLConnection$10.run(HttpURLConnection.java:1945)
	at java.security.AccessController.doPrivileged(Native Method)
	at sun.net.www.protocol.http.HttpURLConnection.getChainedException(HttpURLConnection.java:1944)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1514)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1498)
	at sun.net.www.protocol.https.HttpsURLConnectionImpl.getInputStream(HttpsURLConnectionImpl.java:268)
	at org.kohsuke.github.GitHubHttpUrlConnectionClient$HttpURLConnectionResponseInfo.bodyStream(GitHubHttpUrlConnectionClient.java:197)
	at org.kohsuke.github.GitHubResponse$ResponseInfo.getBodyAsString(GitHubResponse.java:326)
	at org.kohsuke.github.GitHubResponse.parseBody(GitHubResponse.java:91)
	at org.kohsuke.github.Requester.lambda$fetch$1(Requester.java:74)
	at org.kohsuke.github.GitHubClient.createResponse(GitHubClient.java:461)
	at org.kohsuke.github.GitHubClient.sendRequest(GitHubClient.java:412)
	... 12 more
Caused by: java.io.FileNotFoundException: https://api.github.com/repos/apache/spark/statuses/fe0a4b5625ffceaa6c81e9f63637214d08369642
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1896)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1498)
	at java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:480)
	at sun.net.www.protocol.https.HttpsURLConnectionImpl.getResponseCode(HttpsURLConnectionImpl.java:352)
	at org.kohsuke.github.GitHubHttpUrlConnectionClient.getResponseInfo(GitHubHttpUrlConnectionClient.java:69)
	at org.kohsuke.github.GitHubClient.sendRequest(GitHubClient.java:400)
	... 12 more
```

**Comment by HyukjinKwon** [comment]
@ocworld can you take a look https://github.com/apache/spark/pull/32397/checks?check_run_id=2484554770 and enable Github Actions in your forked repository?

**Comment by ocworld** [comment]
@HyukjinKwon Github actions setting is enabled as allow all actions in my forked repository.

**Comment by ocworld** [comment]
@HyukjinKwon 

image

**Comment by HyukjinKwon** [comment]
Can you rebase and force push in this PR to retrigger the build?

**Comment by ocworld** [comment]
@HyukjinKwon I've already rebase and force push. So, now, everything is up-to-date.
After merging others' some pr, i'll retry it.

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 1]
```diff
@@ -308,21 +308,23 @@ private[spark] class SparkSubmit extends Logging {
...
-        // files from a Hadoop Compatible fs e.g. S3. In this case the user might pass:
-        // --packages com.amazonaws:aws-java-sdk:1.7.4:org.apache.hadoop:hadoop-aws:2.7.6
-        if (isKubernetesClusterModeDriver) {
-          val loader = getSubmitClassLoader(sparkConf)
-          for (jar <- resolvedMavenCoordinates) {
-            addJarToClasspath(jar, loader)
-          }
-        } else if (isKubernetesCluster) {
+        if (isKubernetesCluster) {
           // We need this in K8s cluster mode so that we can upload local deps
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:312]:
```suggestion
```

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 2]
```diff
@@ -308,21 +308,23 @@ private[spark] class SparkSubmit extends Logging {
         args.repositories, args.ivyRepoPath, args.ivySettingsPath)
 
       if (resolvedMavenCoordinates.nonEmpty) {
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:310]:
```suggestion
      if (resolvedMavenCoordinates.nonEmpty) {
```

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 3]
```diff
@@ -308,21 +308,23 @@ private[spark] class SparkSubmit extends Logging {
...
-          }
-        } else if (isKubernetesCluster) {
+        if (isKubernetesCluster) {
           // We need this in K8s cluster mode so that we can upload local deps
           // via the k8s application, like in cluster mode driver
           childClasspath ++= resolvedMavenCoordinates
         } else {
+          // In K8s client mode, when in the driver, resolved jars are added in the driver.
+          //
+          // For example we might use the dependencies for downloading
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:318]:
```suggestion
          // For example, we might use the dependencies for downloading
```

**Comment by HyukjinKwon** [comment]
did you do something like https://github.com/apache/spark/pull/32400#issuecomment-831051189? The GitHub Actions should run the build in your forked repository but it seems not.

**Comment by ocworld** [comment]
@HyukjinKwon Oh! you're right.  Workflow in actions tab was not activated. It is enabled now. Thank you.

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 4]
```diff
@@ -308,21 +308,23 @@ private[spark] class SparkSubmit extends Logging {
...
-            addJarToClasspath(jar, loader)
-          }
-        } else if (isKubernetesCluster) {
+        if (isKubernetesCluster) {
           // We need this in K8s cluster mode so that we can upload local deps
           // via the k8s application, like in cluster mode driver
           childClasspath ++= resolvedMavenCoordinates
         } else {
+          // In K8s client mode, when in the driver, resolved jars are added in the driver.
+          //
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:317]:
```suggestion
```

**Comment by mridulm** [comment]
+CC @shardulm94, @xkrogen

**Review by dongjoon-hyun — COMMENTED** [review]
Could you elaborate more about the procedure how you test in the PR description?
```
# How was this patch tested?
In my k8s environment, i tested the code.
```

The best way to protect this feature is having a test coverage at `BasicTestsSuite` or `DepsTestsSuite`. Do you think you can add a new test case test, @ocworld ?

**Comment by holdenk** [comment]
So I'm a little confused about the intention of the PR. If we have a package that we want to resolve in cluster mode we probably do need to put it on compatible filesystem so the workers can fetch the JARs. It's either that or have each worker fetch the resolved JARs and maybe I'm missing it but I don't see the logic for that here.

**Comment by holdenk** [comment]
Ah wait now I'm remembering, we serve the files from the driver to the workers. So yeah the design makes more sense than I thought originally.

+1 on more testing though. We could add a package and then try and load it in the Kube integration tests if doing the manual deps testing is too complicated?

**Comment by ocworld** [comment]
@holdenk Now, I have no idea about checking jars resolved and downloaded in code in kube integrate test. So, it is only tested in my system and my k8s spark env.

**Comment by holdenk** [comment]
I think we really need integration tests for this, would you be up for me to making a PR to your branch to add integration tests?

**Comment by GaruGaru** [comment]
Any update on this ?

**Comment by grcanosa** [comment]
Any news on when this is going to be merged?

**Comment by ocworld** [comment]
> Any news on when this is going to be merged?

It is not merged now. I've tried to add test.

**Comment by rahulbhatia2702** [comment]
is there a solution for this?

**Comment by AlexandrePieroux** [comment]
> is there a solution for this?

In my case I just build the docker images with the dependencies included by hand.

ex:
`RUN curl https://repo1.maven.org/maven2/org/apache/spark/spark-sql-kafka-0-10_2.12/3.1.2/spark-sql-kafka-0-10_2.12-3.1.2.jar --output /opt/spark/jars/spark-sql-kafka-0-10_2.12-3.1.2.jar`

Still looking for the day this is included in spark release.

**Comment by jbguerraz** [comment]
@dongjoon-hyun may be this PR could be re-opened and stale removed since this is still an issue ? Thank you!

**Comment by holdenk** [comment]
The repo that submitted the PR is deleted so we can't re-open it. But feel free to recreate and go for it.

**Comment by ocworld** [comment]
I'll reopen the pr soon.

**Comment by ocworld** [comment]
@jbguerraz @holdenk The pr is recreated in #38828

**Comment by ulan-yisaev** [comment]
> RUN curl https://repo1.maven.org/maven2/org/apache/spark/spark-sql-kafka-0-10_2.12/3.1.2/spark-sql-kafka-0-10_2.12-3.1.2.jar --output /opt/spark/jars/spark-sql-kafka-0-10_2.12-3.1.2.jar

Thank you so much for the workaround, and it's a pity that this bug took two days of my life :)

**Comment by ocworld** [comment]
@jbguerraz @GaruGaru The pr(https://github.com/apache/spark/pull/38828) is merged now

**Comment by jbguerraz** [comment]
Thank you @ocworld :-)

**Pull request closed without merging**